src/array.cr

```diff
@@ -208,6 +208,16 @@ class Array(T)
   def &(other : Array(U)) forall U
     return Array(T).new if self.empty? || other.empty?
 
+    # Heuristic: for small arrays we do a linear scan, which is usually
+    # faster than creating an intermediate Hash.
+    if self.size <= SMALL_ARRAY_SIZE && other.size <= SMALL_ARRAY_SIZE
```

Comment:
Just an idea: `self.size + other.size <= SMALL_ARRAY_SIZE * 2` would match more cases where the total number of comparisons is as low as with the individual limits.
I haven't benchmarked it, though.

Reply:
I think this one is good!